pageserver/src/virtual_file.rs

```diff
@@ -173,55 +173,162 @@ impl OpenFiles {
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub enum CrashsafeOverwriteError {
-    #[error("final path has no parent dir")]
-    FinalPathHasNoParentDir,
-    #[error("remove tempfile: {0}")]
-    RemovePreviousTempfile(#[source] std::io::Error),
-    #[error("create tempfile: {0}")]
-    CreateTempfile(#[source] std::io::Error),
-    #[error("write tempfile: {0}")]
-    WriteContents(#[source] std::io::Error),
-    #[error("sync tempfile: {0}")]
-    SyncTempfile(#[source] std::io::Error),
-    #[error("rename tempfile to final path: {0}")]
-    RenameTempfileToFinalPath(#[source] std::io::Error),
-    #[error("open final path parent dir: {0}")]
-    OpenFinalPathParentDir(#[source] std::io::Error),
-    #[error("sync final path parent dir: {0}")]
-    SyncFinalPathParentDir(#[source] std::io::Error),
+const ENOSPC: i32 = 28;
+const EBADF: i32 = 9;
+
+/// Call this when the local filesystem gives us an error with an external
+/// cause: this includes EIO, EROFS, and EACCESS: all these indicate either
+/// bad storage or bad configuration, and we can't fix that from inside
+/// a running process.
+pub(crate) fn on_fatal_io_error(e: &std::io::Error) {
+    tracing::error!("Fatal I/O error: {}", &e);
+    std::process::abort();
+}
+
+/// Identify error types that should alwways terminate the process.  Other
+/// error types may be elegible for retry.
+pub(crate) fn is_fatal_io_error(e: &std::io::Error) -> bool {
+    match e.kind() {
+        ErrorKind::InvalidInput => {
+            // Invalid input is always a code problem, not a disk/filesystem problem
+            return false;
+        }
+        ErrorKind::UnexpectedEof => {
+            // UnexpectedEof in e.g. read_exact_at tends to indicate a code bug in
```

Comment:
could also be local corruption. maybe it should be mentioned?

Reply:
True, it's one of the places we have ambiguity between code bugs and corruption.

In the latest version of this PR, I have changed the function to only terminate on an explicit list of cases, so this particular ErrorKind isn't explicitly mentioned any more.